pennylane/_qubit_device.py

```diff
@@ -86,6 +86,8 @@ class QubitDevice(Device):
     _imag = staticmethod(np.imag)
     _roll = staticmethod(np.roll)
     _stack = staticmethod(np.stack)
+    _outer = staticmethod(np.outer)
+    _diag = staticmethod(np.diag)
```

Comment:
Just wondering: would it be worth moving this to `default.mixed`? Seems like that would be the only device needing it for now. (Although see the reason of keeping it here too, to have all `staticmethod` defs at one place :thinking: )

Reply:
Note: I just noted a duplication in these staticmethods: `_cast` and `_asarray` are both just `np.asarray`.

Reply:
in numpy yes, but are they the same in tf?

Reply:
> Note: I just noted a duplication in these staticmethods: `_cast` and `_asarray` are both just `np.asarray`.

Yep, this is intentional. In NumPy they are the same function, but in other frameworks they are not; e.g., `tf.cast` vs. `tf.convert_to_tensor`  in TensorFlow, or `torch.astensor` and `lambda x, dtype: x.type(dtype)` in PyTorch.

> 🤔 Thoughts from other reviewers?

It doesn't _really_ need to go here (at the moment). The reason for these attributes is so that the built-in methods of `QubitDevice` (like `QubitDevice.marginal_probs`) can be framework agnostic --- they just call `self._diag` to get the diagonal, rather than relying on NumPy directly. This way, if an inheriting device instead would prefer that TF/Torch/etc are used, they can simply _redefine_ `QubitDevice._diag` to force the parent method to use the framework of choice.

But since nothing in `QubitDevice` is using `self._diag` or `self._outer`, it's not strictly necessary.

But, it's up to you 🙂 Perfectly fine to leave this in if you want; it could make creating a TensorFlow version of the mixed state simulator easier down the line.

Reply:
Thanks all for the clarification! I would prefer to leave here if it makes it easier to create other version of the mixed-state simulator